### PR TITLE
Issue #4682 Session with no attributes unreadable from jdbc

### LINKED ISSUE #4682: "UnreadableSessionDataException Unreadable session ..." after upgrading to 9.4.27
**Jetty version** 9.4.27.v20200227

**Java version** openjdk11

**OS type/version** centos

**Description**

After upgrading from 9.4.25.v20191220 to the above, we often see Warnings like this in our logs: 
`UnreadableSessionDataException
Unreadable session node0q5obwgxy5xig1pk5hkb6lpdbp7 for node0__controlroom_0.0.0.0`

Stacktrace below shows a regular request, but this also occurs when the `Housekeeper` attempts to read session data. The error looks like the `Inputstream` that is passed to the constructor of `ClassLoadingObjectInputStream` is null.

However, I have verified that the table and column do actually contain data:

```sql
select length(map) from jettysessions ;
 length 
--------
   5370
   2804
   4791
   ....
(25 rows)
```

I have dropped and let jetty recreate the table.

Postgres-version is **PostgreSQL 9.3.25**
Jdbc-driver version: **42.2.10** 

Full stacktrace follows:
```
org.eclipse.jetty.server.session.UnreadableSessionDataException: Unreadable session node0q5obwgxy5xig1pk5hkb6lpdbp7 for node0__controlroom_0.0.0.0
    at org.eclipse.jetty.server.session.JDBCSessionDataStore.doLoad(JDBCSessionDataStore.java:640)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.lambda$load$0(AbstractSessionDataStore.java:93)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1367)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1404)
    at org.eclipse.jetty.server.session.SessionContext.run(SessionContext.java:92)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.load(AbstractSessionDataStore.java:101)
    at org.eclipse.jetty.server.session.AbstractSessionCache.loadSession(AbstractSessionCache.java:416)
    at org.eclipse.jetty.server.session.AbstractSessionCache.lambda$getAndEnter$0(AbstractSessionCache.java:353)
    at org.eclipse.jetty.server.session.DefaultSessionCache.lambda$doComputeIfAbsent$0(DefaultSessionCache.java:116)
    at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
    at org.eclipse.jetty.server.session.DefaultSessionCache.doComputeIfAbsent(DefaultSessionCache.java:114)
    at org.eclipse.jetty.server.session.AbstractSessionCache.getAndEnter(AbstractSessionCache.java:346)
    at org.eclipse.jetty.server.session.AbstractSessionCache.get(AbstractSessionCache.java:327)
    at org.eclipse.jetty.server.session.SessionHandler.getSession(SessionHandler.java:906)
    at org.eclipse.jetty.server.session.SessionHandler.getHttpSession(SessionHandler.java:552)
    at org.eclipse.jetty.server.session.SessionHandler.checkRequestedSessionId(SessionHandler.java:1656)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1529)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:717)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
    at org.eclipse.jetty.server.Server.handle(Server.java:500)
    at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
    at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
    at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
    at java.lang.Thread.run(Thread.java:834)
java.io.EOFException: null
    at java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2763)
    at java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3258)
    at java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:873)
    at java.io.ObjectInputStream.<init>(ObjectInputStream.java:350)
    at org.eclipse.jetty.util.ClassLoadingObjectInputStream.<init>(ClassLoadingObjectInputStream.java:50)
    at org.eclipse.jetty.server.session.JDBCSessionDataStore.doLoad(JDBCSessionDataStore.java:634)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.lambda$load$0(AbstractSessionDataStore.java:93)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1367)
    at org.eclipse.jetty.server.handler.ContextHandler.handle(ContextHandler.java:1404)
    at org.eclipse.jetty.server.session.SessionContext.run(SessionContext.java:92)
    at org.eclipse.jetty.server.session.AbstractSessionDataStore.load(AbstractSessionDataStore.java:101)
    at org.eclipse.jetty.server.session.AbstractSessionCache.loadSession(AbstractSessionCache.java:416)
    at org.eclipse.jetty.server.session.AbstractSessionCache.lambda$getAndEnter$0(AbstractSessionCache.java:353)
    at org.eclipse.jetty.server.session.DefaultSessionCache.lambda$doComputeIfAbsent$0(DefaultSessionCache.java:116)
    at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1737)
    at org.eclipse.jetty.server.session.DefaultSessionCache.doComputeIfAbsent(DefaultSessionCache.java:114)
    at org.eclipse.jetty.server.session.AbstractSessionCache.getAndEnter(AbstractSessionCache.java:346)
    at org.eclipse.jetty.server.session.AbstractSessionCache.get(AbstractSessionCache.java:327)
    at org.eclipse.jetty.server.session.SessionHandler.getSession(SessionHandler.java:906)
    at org.eclipse.jetty.server.session.SessionHandler.getHttpSession(SessionHandler.java:552)
    at org.eclipse.jetty.server.session.SessionHandler.checkRequestedSessionId(SessionHandler.java:1656)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1529)
    at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1215)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
    at org.eclipse.jetty.server.handler.gzip.GzipHandler.handle(GzipHandler.java:717)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
    at org.eclipse.jetty.server.Server.handle(Server.java:500)
    at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
    at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:547)
    at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
    at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
    at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
    at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
    at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
    at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
    at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
    at java.lang.Thread.run(Thread.java:834)
```

### PR DESCRIPTION
**Pull request by janbartel**
Closes #4682 

The explanation is in the original bug report. 

This fix ensures that even if the session attributes is empty, we still use the ObjectOutputStream to serialize the empty map, because when reading the session back, the ObjectInputStream needs to see the proper stream headers. Previously, we had tried to save bandwidth by writing 0 bytes, however the ObjectInputStream doesn't recognize that as a valid stream.

Also added a couple of tests for reading/writing/updating empty session attributes.

This bug only affects sessions using jdbc.

### CODE REVIEW

File: tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java [hunk 1]
```diff
@@ -131,6 +130,35 @@ public static void prepareTables() throws SQLException
...
+            String dump = "id=" + id +
+                          " ctxt=" + context +
+                          " node=" + node +
+                          " exp=" + expires +
+                          " acc=" + accessed +
+                          " lacc=" + lastAccessed +
+                          " ck=" + cookieSet +
+                          " lsv=" + lastSaved +
+                          " blob length=" + blob.length();
+            System.err.println(dump);
```
Comment (olamy) [inline tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java:159]: maybe use logging :)

File: tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java [hunk 2]
```diff
@@ -151,6 +179,7 @@ public static boolean existsInSessionTable(String id, boolean verbose)
                 while (result.next())
                 {
                     results = true;
+                    dumpRow(result);
```
Comment (joakime) [inline tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java:182]: Want to leave this System.err.println here?

File: tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java [hunk 3]
```diff
@@ -232,41 +261,53 @@ public static boolean checkSessionPersisted(SessionData data)
...
-        throws Exception
+    
+    public static void insertSession(SessionData data) throws Exception
     {
+
         Class.forName(DRIVER_CLASS);
         try (Connection con = DriverManager.getConnection(DEFAULT_CONNECTION_URL);)
         {
             PreparedStatement statement = con.prepareStatement("insert into " + TABLE +
                 " (" + ID_COL + ", " + CONTEXT_COL + ", virtualHost, " + LAST_NODE_COL +
```
Comment (joakime) [inline tests/test-sessions/test-jdbc-sessions/src/test/java/org/eclipse/jetty/server/session/JdbcTestHelper.java:272]: Any reason why `", virtualHost, "` is not just `", " + VHOST_COL + ", "` ?

**Review by joakime — APPROVED** [review]
LGTM.

I'm happy with it as is, just some minor niggles about dump and missing `VHOST_COL` constant.
